Add optional headless fallback output

Consolatis opened the pull request:
Follow-up from
- #1523

Copy from `docs/environment`
```
##
## This causes a virtual output to be created automatically whenever there
## are no outputs around. This helps for cases like wayvnc so there is always
## an output available to connect to. The name can be chosen freely but there
## must be no duplicate output names, for this reason using VIRTUAL-x or a
## physical connector name like HDMI-A-1 is not recommended as wlroots may
## want to create outputs with those names later on which would then fail.
##

# LABWC_FALLBACK_OUTPUT=NOOP-fallback
```

Things to do:
- [x] split everything into two commits:
  - [x] `store_wlr_output` arg for `virtual_output_add()` and `wlr_output_set_name()` refactor
  - [x] the actual fallback implementation
- [x] remove debug logging when creating / destroying outputs
- [x] add commit body explaining the use-case (copy paste from `docs/environment`?)
- [x] reconfigure support

spl237 commented:
Tested here - works fine for me! Many thanks.

Consolatis commented on src/output.c:
```diff
@@ -683,11 +695,42 @@ do_output_layout_change(struct server *server)
+}
+
 static void
 handle_output_layout_change(struct wl_listener *listener, void *data)
```
Yes, the most important reason is that outputs can be enabled / disabled during runtime which removes them from the layout but does not call into `output_destroy_notify()`.

Temporarily suspending `server->new_output.link` when adding a virtual output is the only option so that we have access to the resulting `wlr_output` before enabling it. We need that for example to call `wlr_output_set_name()` which can't be called after the output has already been enabled. Maybe I should throw that part into a separate PR because it also simplifies `output_add_virtual()` logic in general (e.g. it gets rid of that temporary `server->headless.pending_output_name` char array).

That is also the reason for the new `struct wlr_output **store_wlr_output` argument, otherwise we can't identify the fallback output in `new_output_notify()` (and all handlers like `handle_output_layout_change()` that are called from there, either directly or indirectly).

Consolatis commented on src/output.c:
```diff
@@ -414,6 +412,20 @@ output_init(struct server *server)
+	const char *fallback_output_name = getenv("LABWC_FALLBACK_OUTPUT");
+	if (!string_null_or_empty(fallback_output_name)) {
+		/* FIXME: needs validation of the output name */
+		server->fallback_output.name = xstrdup(fallback_output_name);
```
I am not sure how well that works when we change the env on reconfigure.

Consolatis commented on src/output.c:
```diff
@@ -683,11 +695,42 @@ do_output_layout_change(struct server *server)
+	if (server->fallback_output.name) {
+		/* Prevents unnecessary layout recalculations */
+		server->pending_output_layout_change++;
+
```
I'll give it some thought, handling reconfigure is already on the todo list in the top post.

Consolatis replied:
I moved the whole logic into `output_update_fallback()` which uses `getenv()` directly and is called on init, reconfigure and layout change. The whole permanent state is reduced to a single `server->fallback_wlr_output` variable.

Consolatis commented on src/output.c:
```diff
@@ -895,7 +938,8 @@ handle_output_power_manager_set_mode(struct wl_listener *listener, void *data)
 void
-output_add_virtual(struct server *server, const char *output_name)
+output_add_virtual(struct server *server, const char *output_name,
+		struct wlr_output **store_wlr_output)
```
To have access to it within the new_output_notify handler, the function only returns after that has already been called (either by wlroots itself or by us due to the manual handler disconnect). Suggestions for a short comment that transports the idea?

Consolatis commented on include/labwc.h:
```diff
@@ -337,6 +337,11 @@ struct server {
+
+	struct {
+		char *name;  /* May be NULL */
+		struct wlr_output *output;
```
done

johanmalm commented:
You've convinced that we're technically in the right space, and it is obviously working!

If you're okay to bear with for a bit longer, I have a few suggestions that might help get to something that's easier to understand in the future by people less familiar with the code-base me the likes of us. I have to admit, I've spent more time reviewing this one than I normally do and I'm feeling a bit dumb here. Apologies. Maybe it's just the flow/naming/structure... or simply just me :smile:

So, in no particular order:

1. `server->headless.started` What does it mean? If it's only used by a couple of functions in output.c, could we not make a static variable in that translation unit. It's set to true in server.c, but only just befor a call to an output.c function. I'm guessing it's protecting against something, but it's not clear what that is. Maybe a comment to describe its purpose too.

2. `output_add_virtual()` is only called from two places. One of those passes NULL to that `**` argument, and other one passes `&server->fallback_wlr_output`. Now, that `fallback_wlr_output` variable is only ever used in `output.c` in three functions:

  - output_update_fallback()
  - output_remove_virtual()
  - output_add_virtual() - via a pointer

I guess you could argue that it's a global state, but it just doesn't feel right. How about creating a separate translation unit for this... maybe `output-virtual.c` (for those three functions). That could then contain `static struct wlr_output *fallback;` and `bool headless_started`. It would feel a bit more encapsulated.

3. Suggest covering that suspend-new-output-notify with comments. For example, it makes sense once you know that `wlr_output_set_name()` can't be called after the output has already been enabled.

Consolatis commented:
> If you're okay to bear with for a bit longer

Heh, obviously :)

> I have a few suggestions that might help get to something that's easier to understand in the future by people less familiar with the code-base

I sometimes suck with this so hints how to write / comment the logic so things are easier to understand are obviously always welcome.

> 1. `server->headless.started` What does it mean? If it's only used by a couple of functions in output.c, could we not make a static variable in that translation unit. It's set to true in server.c, but only just before a call to an output.c function.

It is set to true in `server.c` after the backends were started. The next function in there being `output_update_fallback()` is just a coincidence, the `->started` bool is required for all calls to `output_add_virtual()`. The fallback output is merely one user of the API. Sadly wlroots does not seem to provide a public API to ask the backend itself if it has been started or not so I went with the bool instead. When moving things to `output_virtual.c` we could do the same with something like `output_virtual_set_backend_started(true)` which then sets a static var there.

> I'm guessing it's protecting against something, but it's not clear what that is. Maybe a comment to describe its purpose too.

When creating a new headless output, the headless backend in wlroots will either
- emit the new output signal directly (if the backend has already been started at the point where we add a virtual output)
- or do not emit it directly and then emit new output signals for all headless outputs once the backend has been started

As a result we need to restrict our manual call of the new output handler to such cases where the signal \*would\* be emitted immediately (e.g. when the headless backend has already been started). Otherwise we would end up delivering it twice for the same wlr_output.

> 2. `output_add_virtual()` is only called from two places. One of those passes NULL to that `**` argument, and other one passes `&server->fallback_wlr_output`. Now, that `fallback_wlr_output` variable is only ever used in `output.c` in three functions:
> 
> * output_update_fallback()
> * output_remove_virtual()
> * output_add_virtual() - via a pointer
> 
> I guess you could argue that it's a global state, but it just doesn't feel right. How about creating a separate translation unit for this... maybe `output-virtual.c` (for those three functions). That could then contain `static struct wlr_output *fallback;` and `bool headless_started`. It would feel a bit more encapsulated.

That would be an option, yes. I'll give it a try and see how it turns out.

> 3. Suggest covering that suspend-new-output-notify with comments. For example, it makes sense once you know that `wlr_output_set_name()` can't be called after the output has already been enabled.

Will try to think of something but I seem to be pretty bad at writing a coherent explanation that doesn't miss the important bits.

johanmalm commented:
Thanks. Okay - the jigsaw pieces are starting to come together for me. You must be a genius to have worked all this out!! :smile:

Consolatis commented:
Moved to `output-virtual.c` via
- #1592

and squashed + rebased everything in this branch. I tried to add some sensible comment to explain things but suggestions for improvements are very welcome.

Consolatis commented on src/output-virtual.c:
```diff
@@ -93,3 +99,32 @@ output_virtual_remove(struct server *server, const char *output_name)
+
+	if (wl_list_empty(layout_outputs) && !string_null_or_empty(fallback_output_name)) {
+		wlr_log(WLR_DEBUG, "adding fallback output %s", fallback_output_name);
+		assert(!fallback_output);
```
I am actually not sure about this assert.. It might be possible for a user to trigger this by disabling the fallback output via wlr-randr or similar. The wlroots layout might have a handler for that that removes it from the layout. Need to verify.

Consolatis replied:
Replaced the assert by adding it to the condition instead, this should make the logic more robust.

Consolatis commented:
This PR is about ready from my side.

@spl237 Final test would be appreciated.

spl237 commented:
Should get some time to try this tomorrow - will test and report back then.

johanmalm commented:
LGTM. Not tested.

spl237 commented:
Just tried here - works fine for me. Many thanks!

spl237 commented:
Following on from the possible bug I reported on the other thread, this feels timing-related to me. When I first tested this morning, everything seemed fine. Testing again half an hour ago, every time I connected an HDMI after booting without one, the output crashed on connection. And now, I can't seem to repeat it - I'm plugging and unplugging and everything is fine again.

I don't think I have changed anything else, but it may be something else on my system. I'll do some more digging with a clean image to check. Will report back - don't panic for now!

spl237 commented:
OK, have managed to repeat. Sometimes when it gets into the grey screen state, you can fix it by unplugging and reconnecting the HDMI cable. This does feel like some sort of timing hazard to me. Will try to get a log.

spl237 commented:
There's definitely something not right in the logs.

If I boot to the virtual output only, the log is being spammed heavily with lines like this:

```
00:00:18.549 [DEBUG] [types/output/render.c:187] Enabling direct scan-out on output 'NOOP-1' (locks: 0)
00:00:18.549 [DEBUG] [types/output/cursor.c:44] Enabling hardware cursors on output 'NOOP-1' (locks: 0)
00:00:18.555 [DEBUG] [types/output/render.c:187] Disabling direct scan-out on output 'NOOP-1' (locks: 1)
00:00:18.555 [DEBUG] [types/output/cursor.c:44] Disabling hardware cursors on output 'NOOP-1' (locks: 1)
00:00:18.583 [DEBUG] [types/output/render.c:187] Enabling direct scan-out on output 'NOOP-1' (locks: 0)
00:00:18.583 [DEBUG] [types/output/cursor.c:44] Enabling hardware cursors on output 'NOOP-1' (locks: 0)
00:00:18.590 [DEBUG] [types/output/render.c:187] Disabling direct scan-out on output 'NOOP-1' (locks: 1)
00:00:18.590 [DEBUG] [types/output/cursor.c:44] Disabling hardware cursors on output 'NOOP-1' (locks: 1)
00:00:18.618 [DEBUG] [types/output/render.c:187] Enabling direct scan-out on output 'NOOP-1' (locks: 0)
00:00:18.618 [DEBUG] [types/output/cursor.c:44] Enabling hardware cursors on output 'NOOP-1' (locks: 0)
00:00:18.623 [DEBUG] [types/output/render.c:187] Disabling direct scan-out on output 'NOOP-1' (locks: 1)
00:00:18.623 [DEBUG] [types/output/cursor.c:44] Disabling hardware cursors on output 'NOOP-1' (locks: 1)
00:00:18.654 [DEBUG] [types/output/render.c:187] Enabling direct scan-out on output 'NOOP-1' (locks: 0)
00:00:18.654 [DEBUG] [types/output/cursor.c:44] Enabling hardware cursors on output 'NOOP-1' (locks: 0)
00:00:18.654 [DEBUG] [types/output/render.c:187] Disabling direct scan-out on output 'NOOP-1' (locks: 1)
```

At the point where I connect an HDMI and it goes to a grey screen, the log file just gets corrupted completely - it just starts to fill up with bytes which are all zero - looks like something has seriously crashed.

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

johanmalm commented:
Do the problems only occur when `LABWC_FALLBACK_OUTPUT` is set, or do we need to consider reverting the commit until we find the issue.

spl237 commented:
I don't think the problems occur unless you try to boot without an HDMI cable attached, so unless you are trying to use the fallback output, I don't think you will see the problem. So my guess is that there is no need to revert for now.

Consolatis commented:
I'll try to debug this later on a DRM system. The symptoms look very weird though.

Regarding revert, I am easy on that. It should have no effect unless the env var is actually set.

spl237 commented:
I was testing with wayvnc enabled (latest version 0.8), and saw the problem with both TigerVNC and RealVNC clients, so it's not client-specific. But I did always have a client connected - it was the VNC client on which I could see the grey screen, as the HDMI output didn't appear to be producing anything at all.

I don't have reuseOutputMode set, no.

Don't have any time this weekend, but on Monday I will flash a clean image to make sure it isn't something I have broken elsewhere and see if I can find a repeatable way to provoke the problem. We did move to an updated wayvnc a few days ago, so it's not impossible this is an artifact of that which coincidentally only happened to occur after I merged your patch; I reckon I see the problem perhaps 50% of the time.

I'll do some more testing on Monday and see if I can get any better info.

spl237 commented:
OK, on trying again on a clean image here, I cannot now repeat the problem. This is with a different Raspberry Pi and a different monitor - not sure if the hardware is somehow implicated - but I have been connecting and disconnecting and rebooting for ten minutes with a client connected without any sign of the grey screen.

For now, let's put this down to a one-off glitch in my system, but I'll keep testing and report back if it does recur in the future.
